### PR TITLE
2.6.x branch [JENKINS-39355 follow up][JENKINS-40382] Pick up SCM API 2.x

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-39355](https://issues.jenkins-ci.org/browse/JENKINS-39355) [JENKINS-40382](https://issues.jenkins-ci.org/browse/JENKINS-40382)

@reviewbybees (this will also need merging for the 3.x branch)

### CODE REVIEW

**Comment by stephenc** [comment]
@MarkEWaite This will need to have a coordinated release (as well as the corresponding changes for the 3.0.0 line) as we will need to release SCM-API 2.0.1 before this release (and when we release that we will break github-branch-source and bitbucket-branch-source until we get this release).

Would it be ok for me to run the releases of the git plugin against 2.6.x and 3.0.x with these changes (once you've done whatever pre-checks you want to run?) so that I can ensure that the update center is not in mixed state for too long?

File: src/main/java/jenkins/plugins/git/GitSCMFile.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        if (".".equals(name)) {
+            next = this;
+        } else if ("..".equals(name)) {
+            SCMFile parent = parent();
+            next = parent == null ? this : parent;
+        } else {
+            next = new GitSCMFile(fs, this, name);
+        }
+        String restOfPath = path.substring(index + 1);
+        return StringUtils.isBlank(restOfPath) ? next : next.child(restOfPath);
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/GitSCMFile.java]: This is really awkward. Can `SCMFile` not make it a bit more straightforward?

File: src/test/java/jenkins/plugins/git/GitSampleRepoRule.java [hunk 2]
```diff
@@ -49,6 +51,10 @@ public void init() throws Exception {
         git("commit", "--message=init");
     }
 
+    public final boolean mkdirs(String rel) throws IOException {
```
Comment (jglick) [inline src/test/java/jenkins/plugins/git/GitSampleRepoRule.java]: Really belongs in `AbstractSampleDVCSRepoRule` but fine here for now.

File: src/main/java/jenkins/plugins/git/GitSCMFile.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/jenkins/plugins/git/GitSCMFile.java]: 7651366 should do that for you

**Pull request merged**